rfd/0084-azure-backend.md

```diff
@@ -0,0 +1,130 @@
+
+## Why
+
+To support deployments of Teleport in Azure with the same convenience and reliability that's currently available when using managed storage in AWS and GCP.
+
+## Details
+
+### Auth backend
+
+Even though our backend is ultimately just a key-value store, the managed NoSQL database offering in Azure, called Cosmos DB, is not sufficient for our purpose, because our cache propagation model relies on all auth servers having access to a stream of events that replicates all the changes applied to the backend. Cosmos DB has a change feed, but it doesn't contain every change to values and it doesn't have deletions - as such, it's not usable as the backing store for our auth backend (a "full fidelity" change feed has been [in private preview since 2020](https://azure.microsoft.com/en-us/updates/change-feed-with-full-database-operations-for-azure-cosmos-db/)).
+
+As using Postgres as the backend storage for Teleport has been possible (in Preview) since v9.0.4, and Azure offers a managed Postgres service, we will be adding support for Azure AD authentication to the Postgres backend - that's the only change needed to support Azure Postgres in Teleport. As of August 2022, only the "Single Server" flavor of Azure Postgres supports AD authentication and it's limited to Postgres 11, which will be [deprecated in November 2023](https://docs.microsoft.com/en-us/azure/postgresql/single-server/concepts-version-policy#major-version-retirement-policy), but it seems likely that AD authentication will also be supported in the "Flexible Server" offering by then.
+
+### Configuration
+
+This are some options on how the configuration side might look like.
+
+The original SQL Backend RFD imagined something like:
+
+```yaml
+teleport:
+  storage:
+    type: postgres
+    addr: pgservername.postgres.database.azure.com
+    database: dbname_defaulting_to_teleport
+    tls:
+      ca_file: path/to/azure_postgres_trust_roots.pem
+    azure:
+      username: pgusername@pgservername
+```
+
+Alternatively, we could provide additional aliases for the Postgres backend:
+
+```yaml
+teleport:
+  storage:
+    type: azurepostgres
+    addr: pgservername.postgres.database.azure.com
+    username: pgusername@pgservername
+    database: dbname_defaulting_to_teleport
+    tls:
+      ca_file: path/to/azure_postgres_trust_roots.pem
+```
+
+Or we could add some "authentication type" field:
+
+```yaml
+teleport:
+  storage:
+    type: postgres
+    addr: pgservername.postgres.database.azure.com
+    username: pgusername@pgservername
+    database: dbname_defaulting_to_teleport
+    authentication: azure
+    tls:
+      ca_file: path/to/azure_postgres_trust_roots.pem
+```
+
+Other options such as checking for the `.postgres.database.azure.com` suffix in the hostname seem a bit too fragile.
```

Comment:
Agree on this point. I have no preference for any one of the 3 YAML snippets above.

Reply:
The first option makes it clear that only the azure use case requires a username. The other options would require documenting which scenarios need a username and which extract it from the client cert.